### PR TITLE
Mutation counting

### PR DESCRIPTION
**Pull request by Ward9250**
Implementation of counting of different types of mutations.

A nessecery precursor to the evolutionary distance PR, but sizable enough that I've separated into its own PR.

At first (in this PR), only two kinds of mutation counting will be implemented: counting different nucleotides, and counting the number of transitions and number of transversions.

The count_mutations method uses a strict ambiguity test, which discounts a site completely from the computation if either nucleotide is uncertain or a gap i.e. ambiguity codes.

I tried experimenting (hence all the commits!) with making use of all symbols including ambiguities as sometimes it is possible to tell if two ambiguous sites, are different or are say a transition/transversion despite their ambiguity. However because of the complexity of doing this, plus the bias it could introduce in downstream analysis if applied uncritically (for example you can always detect more mutated sites with ambiguity codes, than you can conserved sites), it is not worth doing - if a sequence were that ambiguous, maybe you need better sequence data in the first place ;).

### CODE REVIEW

**Comment by Ward9250** [comment]
Ready for review. If tests pass I'd like to squash and merge asap if there are no objections.

**Comment by Ward9250** [comment]
Tests now pass, I'd like to merge in the morning (UK) if no objections. There's plenty of room for improvement e.g. bitwise computations for 2 bit sequences if they are known. But for now this works well for the distance computation PR.

File: src/var/Var.jl [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+#
+# This file is a part of BioJulia. License is MIT: https://github.com/BioJulia/Bio.jl/blob/master/LICENSE.md
+
+
+
+module Var
+using Bio.Seq
+
+export
+
```
Comment (kdmurray91) [inline src/var/Var.jl]: pedantry: do we need the extra newlines here?

File: src/var/mutation_counting.jl [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+Count the number of differences between two DNA sequences.
+Returns a tuple of the number of sites that are different, and the number of sites
+considered (sites with any ambiguity characters are not considered).
+"""
+function count_mutations{A<:alphabets}(seq1::BioSequence{A},
+    seq2::BioSequence{A},
+    t::Type{Different})
+
+    @assert length(seq1) == length(seq2)
```
Comment (kdmurray91) [inline src/var/mutation_counting.jl]: Should this use `error()` to return a more informative message to the user?

File: src/var/mutation_counting.jl [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+A, T, G, or C, then this function returns true.
+"""
+@inline function is_ambiguous_strict{T<:Nucleotide}(a::T, b::T)
+    return isambiguous(a) || isambiguous(b)
+end
+
+typealias alphabets Union{DNAAlphabet{2}, DNAAlphabet{4}, RNAAlphabet{2}, RNAAlphabet{4}}
+
+"""
+    count_mutations(seq1::BioSequence{A}, seq2::BioSequence{A}, kind::Type{Different})
```
Comment (kdmurray91) [inline src/var/mutation_counting.jl]: is the `{A}` here required? you don't have it in the signatures for the other mutation types.

File: src/var/mutation_counting.jl [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    @assert length(seq1) == length(seq2)
+    nDifferences = 0
+    nSites = length(seq1)
+
+    for (a, b) in zip(seq1, seq2)
+        if is_ambiguous_strict(a, b)
+            nSites -= 1
+            continue
+        end
+        if a != b
```
Comment (kdmurray91) [inline src/var/mutation_counting.jl]: Could one re-factor this to use an if statement on the type of mutation required, to reduce code duplication?

**Comment by kdmurray91** [comment]
Looks great @Ward9250. I've left a couple of comments, but I don't consider them serious enough to block a merge if you need this in master now-ish.

File: docs/src/man/var.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+| :------------- | :-------------------------------------- |
+| `Different`    | Nucleotides that are different          |
+| `Transition`   | Transition mutations                    |
+| `Transversion` | Transversion mutations                  |
+
+`Different` simply compares two nucleotides to see if they can be deemed
+the same or different. Ambiguous cases are detected and excluded from the
+computation.
+
+`TsTv` compares two nucleotides to see if they can be determined
```
Comment (bicycle1885) [inline docs/src/man/var.md]: `TsTv` is unclear to me.

File: docs/src/man/var.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+to be a transition mutation, a transversion mutation, or no mutation at all.
+As with `DifferentBases`, cases where a nucleotide is ambiguous are excluded
```
Comment (bicycle1885) [inline docs/src/man/var.md]: `DifferentBases`: do you mean `Different`?

File: src/seq/nucleotide.jl [hunk 7]
```diff
@@ -123,6 +123,16 @@ function ispyrimidine(nt::Nucleotide)
 end
 
 """
+    isppambiguous(nt::Nucleotide)
+
+Test if nucleotide could be either a Purine or a Pyrimidine.
+"""
+function isppambiguous(nt::Nucleotide)
```
Comment (bicycle1885) [inline src/seq/nucleotide.jl]: Is this really worth defining & exporting from Bio.Seq? There is no abstraction here.

File: src/var/mutation_counting.jl [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+
+# Counting mutations
+# ------------------
+
+# Mutation types
+
+immutable Different end
+immutable Transition end
+immutable Transversion end
```
Comment (bicycle1885) [inline src/var/mutation_counting.jl]: These names are too generic. Names like `TransitionMutation` would be better I think.

File: src/var/mutation_counting.jl [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    is_ambiguous_strict{T<:Nucleotide}(a::T, b::T)
+
+The strictest test of ambiguity, if either nucleotide is not
+A, T, G, or C, then this function returns true.
+"""
+@inline function is_ambiguous_strict{T<:Nucleotide}(a::T, b::T)
+    return isambiguous(a) || isambiguous(b)
+end
+
+typealias alphabets Union{DNAAlphabet{2}, DNAAlphabet{4}, RNAAlphabet{2}, RNAAlphabet{4}}
```
Comment (bicycle1885) [inline src/var/mutation_counting.jl]: All alphabet types are a subtype of `Bio.Seq.Alphabet`. Why do we need this alias?

File: src/var/mutation_counting.jl [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+Returns a tuple of the number of sites that are different, and the number of sites
+considered (sites with any ambiguity characters are not considered).
+"""
+function count_mutations{A<:alphabets}(seq1::BioSequence{A},
+    seq2::BioSequence{A},
+    t::Type{Different})
+
+    @assert length(seq1) == length(seq2)
+    nDifferences = 0
+    nSites = length(seq1)
```
Comment (bicycle1885) [inline src/var/mutation_counting.jl]: `nDifferences` and `nSites` look like CamelCase, which is not a standard way to name variables in Julia, though these are local variables and not a big deal.

File: src/var/mutation_counting.jl [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+end
+
+typealias alphabets Union{DNAAlphabet{2}, DNAAlphabet{4}, RNAAlphabet{2}, RNAAlphabet{4}}
+
+"""
+    count_mutations(seq1::BioSequence{A}, seq2::BioSequence{A}, kind::Type{Different})
+
+Count the number of differences between two DNA sequences.
+Returns a tuple of the number of sites that are different, and the number of sites
+considered (sites with any ambiguity characters are not considered).
```
Comment (bicycle1885) [inline src/var/mutation_counting.jl]: Insert a blank line after the first sentence.

> Include a single one-line sentence describing what the function does or what the object represents after the simplified signature block. If needed, provide more details in a second paragraph, after a blank line.
<http://julia.readthedocs.io/en/latest/manual/documentation/>

**Comment by bicycle1885** [comment]
About the design choice: What about counting all mutations and packing them in a single object like this:
```julia
immutable MutationCounts
    transition::Int
    transversion::Int
    deletion::Int
    insertion::Int
    difference::Int
    length::Int
end

function count_mutations(seq1, seq2)
    ...
    return MutationCounts(...)
end
```

File: src/var/mutation_counting.jl (hunk 8 above)
Comment (Ward9250) [inline src/var/mutation_counting.jl]: Changed names - next commit.

File: src/seq/nucleotide.jl (hunk 7 above)
Comment (Ward9250) [inline src/seq/nucleotide.jl]: I think this is something accidentally left in from when I was trying to be clever and use all the possible ambiguity codes before I decided/realised that doing so was a stupid idea (expanded in first comment). I'll remove this in the next commit.

File: docs/src/man/var.md (hunk 6 above)
Comment (Ward9250) [inline docs/src/man/var.md]: I do, corrected in next commit.

File: docs/src/man/var.md (hunk 5 above)
Comment (Ward9250) [inline docs/src/man/var.md]: Removed in next commit.

File: src/var/mutation_counting.jl (hunk 9 above)
Comment (Ward9250) [inline src/var/mutation_counting.jl]: I had this in there so as to exclude the amino acid, character and void alphabets. This PR only handles counting mutations in DNA and RNA sequences.

File: src/var/mutation_counting.jl (hunk 10 above)
Comment (Ward9250) [inline src/var/mutation_counting.jl]: Changed to non camel in next commit.

File: src/var/mutation_counting.jl (hunk 11 above)
Comment (Ward9250) [inline src/var/mutation_counting.jl]: Added in next commit.

File: src/var/Var.jl (hunk 1 above)
Comment (Ward9250) [inline src/var/Var.jl]: Extra lines under header removed in next commit.

File: src/var/mutation_counting.jl (hunk 4 above)
Comment (Ward9250) [inline src/var/mutation_counting.jl]: I see what you mean. I hesitate to add a run-time decision on the mutation type though. Would a generated function work for this purpose?

File: src/var/mutation_counting.jl (hunk 9 above)
Comment (bicycle1885) [inline src/var/mutation_counting.jl]: OK, in that case, more explicit name like `NucleotideAlphabet` will make sense.

**Comment by Ward9250** [comment]
Thanks for comments, I've responded to each one and changes will be pushed from my desktop in next commit, but for now it's 3:31am so I'm signing off for a bit!

@bicycle1885 Re the design choice. I would indeed like a method that returns a summary of all the mutation counts. However for distance computations only info on one or two mutation types is required e.g. JC69 only considers `Different` mutations, whilst K80 considers `Transition` and `Transversion` mutations (hence my provision of a `count_mutations` methods accepting both of those types to compute both together in one loop). However, if the overhead of computing say `Transitions` and `Transversion` counts, even though you only need `Different` counts is negligible then it might be a better idea.

Actually I think one method that computes them all is a good idea anyway, but the above was my reasoning behind individual methods as well.

File: src/var/mutation_counting.jl (hunk 9 above)
Comment (Ward9250) [inline src/var/mutation_counting.jl]: Awesome, I've added the name change and will be in next commit!

**Comment by Ward9250** [comment]
Ok, just completed a rebase to squash, when check turn green I'll merge.

**Comment by Ward9250** [comment]
There is an error on OSX, but I believe it is to do with HDF5, and not related to the code submitted in this PR.

**Pull request merged**